core/src/main/clojure/xtdb/metadata.clj

```diff
@@ -270,8 +268,56 @@
             (write-col-meta! true col))
           (.endList cols-wtr))))))
 
+(defn ->table-metadata-idxs [^IVectorReader metadata-rdr]
+  (let [page-idx-cache (HashMap.)
+        meta-row-count (.valueCount metadata-rdr)
+        ^IVectorReader #_<Int> page-idx-rdr (.structKeyReader metadata-rdr "page-idx")
+        ^IVectorReader #_<List> cols-rdr(.structKeyReader metadata-rdr "columns")
+        ^IVectorReader col-rdr (.listElementReader cols-rdr)
+        ^IVectorReader column-name-rdr (.structKeyReader col-rdr "col-name")
+        ^IVectorReader root-col-rdr (.structKeyReader col-rdr "root-col?")
+        col-names (HashSet.)]
+
+    (dotimes [meta-idx meta-row-count]
+      (let [cols-start-idx (.getListStartIndex cols-rdr meta-idx)
+            page-idx (if-let [page-idx (.getObject page-idx-rdr meta-idx)]
+                       page-idx
+                       -1)]
+        (dotimes [cols-data-idx (.getListCount cols-rdr meta-idx)]
+          (let [cols-data-idx (+ cols-start-idx cols-data-idx)
+                col-name (str (.getObject column-name-rdr cols-data-idx))]
+            (.add col-names col-name)
+            (when (.getBoolean root-col-rdr cols-data-idx)
+              (.put page-idx-cache [col-name page-idx] cols-data-idx))))))
+
+    {:col-names (into #{} col-names)
+     :page-idx-cache (into {} page-idx-cache)
+     :page-count (loop [page-count 0, idx 0]
+                   (cond
+                     (>= idx meta-row-count) (inc page-count)
+                     :else (recur (cond-> page-count
+                                    (not (nil? (.getObject page-idx-rdr idx))) inc)
+                                  (inc idx))))}))
+
+(defn- table-name->dir [table-name] (format "tables/%s/chunks" table-name))
+
+(defn ->table-leaf-obj-key [table-name chunk-idx]
+  (format "%s/leaf-c%s.arrow" (table-name->dir table-name) chunk-idx))
+
+(defn ->table-trie-obj-key [table-name chunk-idx]
+  (format "%s/trie-c%s.arrow" (table-name->dir table-name) chunk-idx))
```

Comment:
Would have preferred for the above 3 to stay in `xtdb.trie` but that would have created a cyclic dep.